src/widget/textbox.rs

```diff
@@ -133,34 +133,58 @@ impl Widget for TextBox {
 
     fn layout(
         &mut self,
-        bc: &BoxConstraints,
-        _children: &[Id],
-        _size: Option<Size>,
-        _ctx: &mut LayoutCtx,
-    ) -> LayoutResult {
-        LayoutResult::Size(bc.constrain((self.width, BOX_HEIGHT)))
-    }
-
-    fn mouse(&mut self, event: &MouseEvent, ctx: &mut HandlerCtx) -> bool {
-        if event.count > 0 {
-            ctx.set_focused(true);
-            ctx.invalidate();
-        }
-        true
+        _layout_ctx: &mut LayoutCtx,
+        _bc: &BoxConstraints,
+        _data: &String,
+        _env: &Env,
+    ) -> Size {
+        Size::new(self.width, BOX_HEIGHT)
     }
 
-    fn key_down(&mut self, event: &KeyEvent, ctx: &mut HandlerCtx) -> bool {
+    fn event(
+        &mut self,
+        event: &Event,
+        ctx: &mut EventCtx,
+        data: &mut String,
+        _env: &Env,
+    ) -> Option<Action> {
         match event {
-            event if event.key_code == KeyCode::Backspace => {
-                self.text.pop();
+            Event::MouseDown(_) => {
+                if ctx.is_hot() {
+                    ctx.set_active(true);
+                } else {
+                    ctx.set_active(false);
+                }
+                ctx.invalidate();
             }
-            event if event.key_code.is_printable() => {
-                self.text.push_str(event.text().unwrap_or(""))
+            Event::KeyDown(key_event) => {
+                match key_event {
+                    event if event.key_code == KeyCode::Backspace => {
+                        let mut text = data.clone();
```

Comment:
I think we can just do `data.pop()`? I forget exactly how this works, but I think that the state that is passed into `event` can be mutated directly, although we might want this to be an `Arc<String>`. @raphlinus can clarify?

(It likely goes without saying that the actual logic for what backspace does is complicated, and is definitely not just popping the last unicode scalar, but this is okay for now)

Reply:
Yes, we should be able to do a `data.pop()` here. I have the `clone` dance in places where data has interior mutability (in some cases, `make_mut` might be better, but we might want to experiment, as refcount == 1 might be rare in practice).

Whether this should be `String` or `Arc<String>` is a pretty complex question. If it's used for relatively small fields I don't think it matters much and would lean toward the former for ergonomics. A general heuristic is that any time you get a chunk of data big enough you don't want to compare it when processing `update`, wrap it in `Arc`.

Regarding backspace, there is a big discussion to be had about this. I have on my local machine an embryonic "xi-text-core" crate that has the backspace logic factored out, and I know @cmyr has also been exploring this. I'd like to do something like that, but there are some really big decisions to be made; for example, this code currently depends on xi-rope, and that seems possibly too heavyweight for this. I definitely don't want to block basic text widget functionality on figuring out those big questions.